src/whisper.cpp

```diff
@@ -5647,6 +5648,35 @@ int whisper_full_with_state(
                     return -8;
                 }
 
+                // Calculate no_speech probability after first decode
+                {
+                    const float * logits = state->logits.data();
+                    const int n_vocab = ctx->vocab.n_vocab;
+
+                    // Find max element for numerical stability
+                    float max_logit = -INFINITY;
+                    for (int i = 0; i < n_vocab; ++i) {
+                        max_logit = std::max(max_logit, logits[i]);
+                    }
+
+                    // Calculate softmax
+                    float sum_exp = 0.0f;
+                    std::vector<float> probs(n_vocab);
+                    for (int i = 0; i < n_vocab; ++i) {
+                        float exp_val = expf(logits[i] - max_logit);
+                        sum_exp += exp_val;
+                        probs[i] = exp_val;
+                    }
+
+                    // Normalize
+                    for (int i = 0; i < n_vocab; ++i) {
+                        probs[i] /= sum_exp;
+                    }
+
+                    // Get probability of no_speech token
+                    state->no_speech_prob = probs[whisper_token_nosp(ctx)];
+                }
+
```

Comment:
This likely has to be done inside `whisper_process_logits` in order to avoid computing the softmax again just for this probability.

Reply:
Unfortunately we cannot reuse the softmax computed inside `whisper_process_logits` since `no_speech_prob` has to be calculated before any logits filtering. Otherwise we get some wrong `no_speech_prob` values. The same method is followed in openai's whisper as well. https://github.com/openai/whisper/blob/main/whisper/decoding.py#L689-L703
Since this `no_speech_prob` calculation is only for the first token in the sequence, it will not cause a big performance impact.
On a related note, I have now modularized the probs calculation and now reusing the same code as `whisper_process_logits`